lib/qa.rb

```diff
@@ -106,18 +106,18 @@ def ensure_qa_enabled
   end
 
   def ensure_can_act
-    if Topic.voted(@post.topic, @user)
-      if self.action_name === QuestionAnswer::Vote::CREATE
-        raise Discourse::InvalidAccess.new, I18n.t('vote.error.alread_voted')
-      end
 
-      if self.action_name === QuestionAnswer::Vote::DESTROY && !QuestionAnswer::Vote.can_undo(@post, @user)
-        raise Discourse::InvalidAccess.new, I18n.t('vote.error.undo_vote_action_window',
-          minutes: SiteSetting.qa_undo_vote_action_window
-        )
-      end
-    elsif self.action_name === QuestionAnswer::Vote::DESTROY
-      raise Discourse::InvalidAccess.new, I18n.t('vote.error.user_has_not_voted')
+    if Topic.can_vote(@post.topic, @user)
```

Comment:
The logic inside this conditional doesn't quite work. The ``can_vote`` check will only return true if the user is under their vote limit. The first conditional will never be true and the second is does not consider the case when a user is exactly at their vote limit (they should still be able to remove a vote then).

For context, the purpose of the method ``ensure_can_act`` is to ensure that a person is able / allowed to either create a vote or destroy a vote. If you look at the top of the class you'll see that it's run before each create and destroy action in the controller:

```
before_action :ensure_can_act, only: [:create, :destroy]
```

To simplifiy this, let's actually get rid of the ``ensure_can_act`` method (and the line quoted above) and move the checks it's performing to the ``create`` and ``destroy`` methods in the class.

Reply:
okay I'll post notes here as I work, and I'll let you know when I've pushed and this second round of updates is ready to review! I've removed the `ensure_can_vote` and I really like the suggestion to move into CREATE and DESTROY - that way we don't need to do (what I thought) were checks for the specific view. So what I basically have is to raise an exception in create if the user is not under the vote limit (meaning they are at it, and unless we've messed something up, they shouldn't be over it). And then in destroy, the only limit is having zero votes (nothing to destroy).

```ruby
  def create

    if !Topic.can_vote(@post.topic, @user)
        raise Discourse::InvalidAccess.new, I18n.t('vote.error.user_over_limit')
    end

    if QuestionAnswer::Vote.vote(@post, @user, vote_args)
      render json: success_json.merge(
        vote_count: @post.vote_count
      )
    else
      render json: failed_json, status: 422
    end
  end

  def destroy

    if Topic.vote_count(@post.topic, @user) == 0
       raise Discourse::InvalidAccess.new, I18n.t('vote.error.user_has_not_voted')
    end

    if QuestionAnswer::Vote.vote(@post, @user, vote_args)
      render json: success_json.merge(
        vote_count: @post.vote_count
      )
    else
      render json: failed_json, status: 422
    end
  end
```
And the `before_action` line is removed. Ok, moving on to next comment.